File: safehere_flutterapp/lib/models/emergency_event.dart
class LocationPoint {
  final double latitude;
  final double longitude;
  final String timestamp;

  LocationPoint({required this.latitude, required this.longitude, required this.timestamp});

  factory LocationPoint.fromJson(Map<String, dynamic> json) => LocationPoint(
        latitude: (json['latitude'] as num).toDouble(),
        longitude: (json['longitude'] as num).toDouble(),
        timestamp: json['timestamp'] ?? '',
      );
}

class EmergencyEvent {
  final String id;
  final String ownerId;
  final String type;
  final String status; // active | resolved | cancelled | false_alarm
  final double? latitude;
  final double? longitude;
  final List<LocationPoint> locationHistory;
  final int? batteryLevel;
  final String startedAt;
  final String? resolvedAt;
  final int contactsNotified;
  final bool evidenceRecording;
  final int evidenceUploadProgress;
  final List<String> evidencePhotos;
  final List<String> evidenceVideos;
  final List<String> evidenceClips;

  EmergencyEvent({
    required this.id,
    required this.ownerId,
    required this.type,
    required this.status,
    this.latitude,
    this.longitude,
    required this.locationHistory,
    this.batteryLevel,
    required this.startedAt,
    this.resolvedAt,
    required this.contactsNotified,
    required this.evidenceRecording,
    required this.evidenceUploadProgress,
    required this.evidencePhotos,
    required this.evidenceVideos,
    required this.evidenceClips,
  });

  bool get isActive => status == 'active';

  factory EmergencyEvent.fromJson(Map<String, dynamic> json) => EmergencyEvent(
        id: json['id'],
        ownerId: json['owner_id'],
        type: json['type'],
        status: json['status'],
        latitude: (json['latitude'] as num?)?.toDouble(),
        longitude: (json['longitude'] as num?)?.toDouble(),
        locationHistory: (json['location_history'] as List? ?? [])
            .map((e) => LocationPoint.fromJson(e))
            .toList(),
        batteryLevel: json['battery_level'],
        startedAt: json['started_at'],
        resolvedAt: json['resolved_at'],
        contactsNotified: json['contacts_notified'] ?? 0,
        evidenceRecording: json['evidence_recording'] ?? false,
        evidenceUploadProgress: json['evidence_upload_progress'] ?? 0,
        evidencePhotos: List<String>.from(json['evidence_photos'] ?? []),
        evidenceVideos: List<String>.from(json['evidence_videos'] ?? []),
        evidenceClips: List<String>.from(json['evidence_clips'] ?? []),
      );
}

File: backend_safeher/app/models/emergency_event.py
import uuid
from datetime import datetime, timezone
import enum

from beanie import Document
from pydantic import Field
from pymongo import IndexModel


def _uuid() -> str:
    return str(uuid.uuid4())


class EmergencyType(str, enum.Enum):
    sos = "sos"
    fall_detected = "fall_detected"
    voice_distress = "voice_distress"
    manual = "manual"
    route_deviation = "route_deviation"


class EmergencyStatus(str, enum.Enum):
    active = "active"
    resolved = "resolved"
    cancelled = "cancelled"
    false_alarm = "false_alarm"


class EmergencyEvent(Document):
    id: str = Field(default_factory=_uuid)
    owner_id: str

    type: EmergencyType = EmergencyType.sos
    status: EmergencyStatus = EmergencyStatus.active

    location_label: str | None = None
    latitude: float | None = None
    longitude: float | None = None

    # list[{latitude, longitude, timestamp}]
    location_history: list = Field(default_factory=list)

    battery_level: int | None = None
    started_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    resolved_at: datetime | None = None

    contacts_notified: int = 0
    evidence_recording: bool = True
    evidence_upload_progress: int = 0

    # Google Drive evidence folder — created immediately on SOS activation.
    # drive_link_sent tracks whether trusted contacts have already received
    # the folder link (it must only be sent once, per the spec).
    drive_folder_id: str | None = None
    drive_folder_link: str | None = None
    drive_link_sent: bool = False

    # stored as relative file paths under MEDIA_ROOT, e.g. "photos/<id>/front_....jpg"
    evidence_photos: list = Field(default_factory=list)
    evidence_videos: list = Field(default_factory=list)
    evidence_clips: list = Field(default_factory=list)  # audio

    class Settings:
        name = "emergency_events"
        indexes = [
            IndexModel("owner_id"),
        ]
